### ISSUE TITLE
Async Router Guard Navigation Issue

### ISSUE THREAD

**Issue by patrickhousley** [angular/angular#16057]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When the URL is changed manually to a parent route that redirects to a child route and that child route navigates to different route with an async guard, the last child route does not always load.

**Expected behavior**
The last child route should load as expected.

**Minimal reproduction of the problem with instructions**
Plunker: http://plnkr.co/edit/PNubvwfYsOTkwPJHrsPb?p=preview
If you remove the guard, everything works as expected.

**Steps to reproduce**
1. Open the plunker
2. Open the execution window in it's own window
3. Copy the URL without the hash (ex. http://run.plnkr.co/mOTJamm7zWdPs0N2/)
4. Paste the URL in a new tab of your browser. This is necessary so you have access to change the hash.
5. Let the page load normally once. You will be directed to `#/test => #/test/testStart => #/test/123`
6. Delete the ID portion of the URL leaving `#/test`. Press enter.
7. You should see in the console the router guard resolves but the route never changes.

** Additional Info**
These are some screenshots from an actual application. I added some debug code to the router. I hope this helps. The first image is a result of me manually modifying the hash in the URL to a parent route. The second is the first child route navigating to a second child route which contains an async guard. Not sure what the third one is or where it came from. There are no additional calls to navigate or navigateByUrl.

<img width="744" alt="screen shot 2017-04-17 at 12 53 44" src="https://cloud.githubusercontent.com/assets/3984979/25101618/9289671e-237a-11e7-9f6f-1119a78f5883.png">
<img width="749" alt="screen shot 2017-04-17 at 12 53 58" src="https://cloud.githubusercontent.com/assets/3984979/25101623/94939bf6-237a-11e7-9219-5872792b0d70.png">
<img width="745" alt="screen shot 2017-04-17 at 12 54 06" src="https://cloud.githubusercontent.com/assets/3984979/25101629/981c50d8-237a-11e7-9ff7-ffc61625d5a8.png">

My routes configuration (this is a subset of the entire app):
```
export const CHANGE_SERVICES_PLANS_ROUTES: Routes = [
  {
    path: 'change-services',
    children: [
      {
        path: '',
        redirectTo: 'select-device',
        pathMatch: 'full'
      },
      {
        path: 'select-device',
        component: ChangeServicesSelectDeviceComponent,
        canActivate: [
          KMSISessionGuard
        ]
      },
      {
        path: 'error',
        component: ChangeServicesPlansErrorPageComponent
      },
      {
        path: ':subscriptionId',
        component: MakeChangesComponent,
        canActivate: [
          KMSISessionGuard,
          ChangeServicesPlansGuard
        ]
      },
      {
        path: ':subscriptionId/review-changes',
        component: ReviewChangesComponent,
        canActivate: [
          KMSISessionGuard,
          ChangeServicesPlansGuard
        ]
      }
    ]
  },
  {
    path: 'change-plans',
    children: [
      {
        path: '',
        redirectTo: 'select-plan',
        pathMatch: 'full'
      },
      {
        path: 'select-plan',
        component: ChangePlansComponent
      }
    ]
  }
];
```

**Please tell us about your environment:**
Mac OS X 10.11.6
Webstorm 2017.1

* **Angular version:**
4.0.2

* **Browser:**
Chrome 57.0.2987.133
 
* **Language:**
Typescript 2.2.2

* **Node (for AoT issues):**
6.9.2

**Comment by patrickhousley** [comment]
Re-tested with Angular 4.3.1 and issue is still present. Is there any update on this issue?

**Comment by patrickhousley** [comment]
Re-tested with Angular 4.4.0-rc.0 and the issue is still present. Is there any update on this issue?

**Comment by patrickhousley** [comment]
I setup a better plunker for this issue. http://plnkr.co/edit/CGHEV5
Once the plunker runs the first time, the last page will provide a button allowing you to go back a page. Click it and watch the console for the router tracing output. The guard check ends and then the routing is cancelled with no reason given.

**Comment by patrickhousley** [comment]
The code [here](https://github.com/angular/angular/blob/66f0ab0371c22af90c9526473769da18b4a95a62/packages/router/src/router.ts#L696) is getting triggered because `appliedUrl`, `state`, `shouldActivate` are all undefined.

**Comment by patrickhousley** [comment]
Using the router to navigate resolves this issue as does enclosing the `location.back()` call in a `setTimeout`. However, this is still a very concerning bug because using the browser's back button effectively reproduces the same scenario as the `location.back()`.

**Comment by jasonaden** [comment]
I think this is a duplicate of #17023. We've got some issues around redirecting in the router and are working on fixes. These typically manifest themselves by the router behaving differently when you deep link (or manually enter a URL) and when you navigate in an already running application.

Please take a look at the linked issue. It's not specific to Upgrade apps, but exists in Angular-only apps as well.

**Comment by patrickhousley** [comment]
To be honest @jasonaden, I don't think it matters one way or the other. Either one or both could be open. It would be nice to see some activity from the Angular team. Even just an acknowledgment of either issue would be nice. No offense intended to anyone.

**Comment by patrickhousley** [comment]
@jasonaden I was wondering if there were any updates on these router issues? Possibly targeting 6.0?

**Comment by patrickhousley** [comment]
Updated the [plunkr](http://embed.plnkr.co/CGHEV5/) to use angular 6 beta.5.

**Comment by jasonaden** [comment]
@patrickhousley Thank you for keeping the plunker up-to-date on this. I will increase the priority on this issue, but to be honest I'm not sure when we will be able to pull it in. But with the increased priority, it will be one of the first that gets looked at when issues are being added to current work.

**Comment by patrickhousley** [comment]
@jasonaden et al.,

Any update on this issue? I have updated the [plunkr](https://embed.plnkr.co/CGHEV5/) again.

**Comment by jasonaden** [comment]
@patrickhousley It looks like I might be addressing this in some work I'm doing right now. You could keep an eye on #24618 and PR #25740 which is laying the groundwork for the aforementioned issue (plus a few others). I believe this issue might be solved by the PR #25740 since we're cleaning up the observable stream in the router and this should help significantly with many of the redirect issues that are floating around.

**Comment by patrickhousley** [comment]
@jasonaden no, the issue is not resolved. I have updated the [plunkr](https://embed.plnkr.co/CGHEV5/) again to Angular 7.1.3.

**Comment by atscott** [comment]
Confirmed, this is still an issue in v9. Reproduction in stackblitz: https://stackblitz.com/edit/angular-ivy-hhg9hj?file=src%2Findex.html

**Comment by atscott** [comment]
Closing, this appears to have been resolved with changes to the Router since v9.
